chore(express): add architecture readme for lightning (#5447)
BitGo/BitGoJS: modules/express/src/lightning/README.md

* TLS: We highly recommend using the `signer lnd` node to create the TLS certificate. If none
is provided during startup, it will automatically create one. The TLS certificate must account
for the IP addresses or hostnames that you will be contacting the `signer lnd` node from. You
can add these in the [LND config](https://github.com/lightningnetwork/lnd/blob/master/sample-lnd.conf) with `tlsextraip` and `tlsextradomain`, respectively.

## Wallet Creation

Unlike on-chain Bitcoin Wallets, the Lightning wallet requires multiple steps to initialize. Broadly, we need to
initiate the deployment of the watch-only node in our infrastructure, configure express and the remote signer with
the IP information of the watch-only node, and finally initialize the lightning wallet with the wallet passphrase.

#### Step (0) Add Lightning Wallet license to your enterprise

Reach out to `support@bitgo.com` to request a license for you to create lightning wallets.

#### Step (1): Initiate Wallet Creation

To initiate wallet creation, use the standard endpoint for wallet creation with the following parameters:

```
POST /api/v2/[t]lnbtc/wallet/generate
{
  label,
  passphrase,
  enterprise,

>   passcodeEncryptionCode

Author: do we actually need this?

following endpoints:

```
GET /api/v2/ping (ping Express)
GET /api/v2/[t]lnbtc/wallet/:walletId/state (ping signer lnd via Express)
```

Both should return `200`. If `/api/v2/ping` is not working, you cannot correct connect to Express. If `/api/v2/[t]lnbtc/wallet/:walletId/state` is not working, then `Express` cannot properly connect to the the `signer lnd`.

#### Step (3) Initialize the `signer lnd` node

Once we are able to communicate with the `signer lnd` node from Express, we can initialize the `signer lnd` node with the following endpoint:

```
POST /api/v2/[t]lnbtc/wallet/:walletId/initwallet
{
  passphrase: string
}
```

This will initialize the lnd using the [`initwallet`](https://lightning.engineering/api-docs/api/lnd/wallet-unlocker/init-wallet/) LND route. This will create the admin macaroon for the signer node and an encrypted version (to the wallet passphrase) will be uploaded to the wallet.

#### Step (4) Update Wallet with relevant information

Once we have a deployment for `Express` and `lnd signer` node, we can update the wallet:

```
PUT /api/v2/[t]lnbtc/wallet/:walletId
{
  signerHost: string,

>   tlsCert: string,

Author: Ideally this would come from the config itself and the customer wouldnt have to provide that here again

#### Step (5) Wait for the `watchOnlyIp` to be set

In the background, initiated by step (1), BitGo is creating the `watch-only lnd` node in our
own infrastructure, and therefore can take some time. You can periodically poll the wallet
with `GET /api/v2/[t]lnbtc/wallet/:walletId` and look for `watchOnlyIp` to be set in the
`coinSpecific` of the wallet. However, we will send you an email once it is set so that you
do not have to.

#### Step (6) Create the signer macaroon

Once the `watchOnlyIp` of the wallet is set, we now are able to create the signer macaroon.
This macaroon will be used by the `watch-only lnd` node to tell the `signer lnd` node to do
certain operational actions such as creating lightning channels, signing invoices, etc, and
is derived from the admin macaroon of the `signer lnd` node. We need the `watchOnlyIp` of the
`watch-only lnd` node so that we can add the caveat to the signer macaroon that it can only
be used by that IP address.

To create the signer macaroon use the following endpoint:

```
POST /api/v2/[t]lnbtc/wallet/:walletId/signermacaroon
{

>   passphrase

Author: https://bitgoinc.atlassian.net/browse/BTC-1871

|  | watch-only lnd +-----------------+----+---+-->| signer lnd node |  |   |
|  +-------+----+---+                 |    |   |   +--------+--------+  |   |
|          |    ^                     |    |   |            ^           |   |
|          |    |                     |    |   |            |           |   |
|          |    |                     |    |   +------------+-----------+   |
|          |    |                     |    |                |               |
|          |    |       +----------+  |    |           +----+----+          |
|          |    |       |          |  |    |           |         |          |
|          |    +------>|  BitGo   |<-+----+-----------+ Express |          |
|          |            | Platform |  |    |           |         |          |
|          |            |          |  |    |           +---------+          |
|  +-------v--------+   +----------+  |    +--------------------------------+
|  | bitcoind/btcd  |                 |
|  +----------------+                 |
+-------------------------------------+
```

[Adapted from this figure.](https://github.com/lightningnetwork/lnd/blob/master/docs/remote-signing.md#remote-signing)

> * `Lightning p2p network`: This is the peer-to-peer lightning network where off-chain settlements are performed. To connect
> to the lightning network, we (BitGo) run a `watch-only` node that acts as the interface between the lightning network and
> your wallet
> * `watch-only lnd`: This is a LND node that contains no private information. It is initialized with the public keys that are
> created when you generate a wallet. However, the private keys stay with you, running along side your express in the
> `signer lnd node`, where they are offline except for a few very specific connections. The `watch-only lnd` node's job is to
> communicate what is happening on the lightning network with BitGo and your signing node, but nothing else. It contains
> no ability or permissions to move funds.

Reviewer: ```suggestion
* `Lightning p2p network`: This is the peer-to-peer lightning network that performs off-chain settlements. To connect to the lightning network, BitGo runs a `watch-only` node that acts as the interface between the lightning network and your wallet.
* `watch-only lnd`: This is an LND node that doesn't contain any private information. It initializes using the public keys that you created when you generated the wallet. Meanwhile, you manage your private keys using Express in the `signer lnd node`. Your private keys remain offline, except for a few specific connections. The `watch-only lnd` node communicates what's happening on the lightning network to BitGo and your signing node - nothing else. The `watch-only lnd` doesn't contain any ability or permissions to move bitcoin.
```

>   The node is also connected to the bitcoin network (`bitcoind/btcd`), where the node watches for funds moving on/off the
>   wallet and to watch for any closing/opening of lightning channels.
> * `signer lnd node`: This is the node that contains the private key information used to authorize the transfer for funds,
> both on-chain and off-chain. This node is not connected to the lighting network or bitcoin directly. Instead, information
> about what is happening comes from the `watch-only lnd` node. You can think of this node being the other half for the LND
> node, and it is completely in the customers control. This signing node is deployed within your own infrastructure along
> side of the `Express` app. Only `Express` and the `watch-only lnd` node can make connections to this application. We
> separate the `watch-only` and `signer` LND nodes because:

Reviewer: ```suggestion
The node also connects to the Bitcoin network (`bitcoind/btcd`). Here, the node watches for bitcoin moving in or out of the wallet and also watches for any lightning channels opening or closing.
* `signer lnd node`: This node contains your private key information that you use to authorize transfers, both on-chain and off-chain. This node isn't connected to the lighting network or Bitcoin directly. Instead, information about what's happening comes from the `watch-only lnd` node. The `signer lnd node` node is the other half of the LND node and is completely in your control. This signing node is deployed within your infrastructure alongside the `Express` app. Only `Express` and the `watch-only lnd` node can make connections to this application. BitGo separates the `watch-only` and `signer` LND nodes because:
```

>   1. It is safer; the private keys are not directly online.
>   2. Keeps the keys in your control while delegating most of the operational complexities of running LND nodes to BitGo.
> * `Express`: This is the self-hosted server that you are running that makes requests to BitGo and the `signer lnd node`.
> * `BitGo Platform`: This is a server that we run to serve the requests of customers. For more information, see the
> [developer documentation](https://developers.bitgo.com/guides/get-started/intro).
> * `bitcoind/btcd`: This is the Bitcoin network.
> * `BitGo Infrastructure`: These are services and connections that are managed by BitGo. You as a client do not need to
> worry about these services
> * `Client Infrastructure`: These are the services that are going to be deployed in your infrastructure.

Reviewer: ```suggestion
  1. It's safer, because the private keys aren't directly online.
  2. It keeps the keys in your control, while delegating most of the operational complexities of running LND nodes to BitGo.
  
* `Express`: This is the self-hosted server that you run and use to make requests to BitGo and the `signer lnd node`.
* `BitGo Platform`: This is a server that BitGo runs to serve your requests. For more information, see the
[Introduction to BitGo](https://developers.bitgo.com/guides/get-started/intro) on the [Developer Portal](https://developers.bitgo.com/).
* `bitcoind/btcd`: This is the Bitcoin network.
* `BitGo Infrastructure`: These are services and connections that BitGo manages. You don't need to manage these services.
* `Client Infrastructure`: These are the services that deploy in your infrastructure.
```

> We have built our self-custodial lightning product to:
> 
> 1. maximize the control of the signer lnd node while keeping it offline
> 2. minimize the control of the watch-only lnd

Reviewer: ```suggestion
BitGo built self-custody lightning wallets to:

1. Maximize your control of the signer LND node while keeping it offline.
2. Minimize your control of the watch-only LND node.
```
Confirm this is still accurate.
